File: packages/core/rust/benchmarks/check-markdown-downstream.py
#!/usr/bin/env python3
"""Compare a private downstream Markdown snapshot with the experimental probe."""

import collections
import hashlib
import json
import re
import subprocess
import sys


def sha256(path):
    digest = hashlib.sha256()
    with open(path, "rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def features(case):
    source = case["markdown"]
    html = case["php"]
    checks = {
        "obfuscated_link": bool(re.search(r"#\[[^]]+\]\(", source)),
        "notice": "> [!" in source,
        "image": "![" in source,
        "phone": 'data-rot="gry:' in html,
        "email": "class=nojs" in html,
        "date": "date(" in source,
        "attributes": bool(re.search(r"\{[.#]|\{(?:id|class)=", source)),
    }
    return [name for name, present in checks.items() if present] or ["unclassified"]


def compare(binary, cases, counts, page_equal, supported_only):
    if not cases:
        return
    request = json.dumps(
        {
            "documents": [case["markdown"] for case in cases],
            "fenced_code_pre_class": cases[0]["pre_class"],
            "supported_only": supported_only,
        },
        ensure_ascii=False,
    )
    result = subprocess.run([binary], input=request, text=True, capture_output=True, check=True)
    output = json.loads(result.stdout)
    if len(output) != len(cases):
        raise RuntimeError("Probe returned a different document count")
    for case, actual in zip(cases, output):
        counts["blocks"] += 1
        page = case["page"]
        page_equal.setdefault(page, True)
        if actual is None and supported_only:
            counts["declined_to_php"] += 1
            continue
        if actual == case["php"]:
            counts["exact"] += 1
            if supported_only:
                counts["accepted_native"] += 1
        else:
            counts["mismatch"] += 1
            page_equal[page] = False
            names = features(case)
            if any(name in {"obfuscated_link", "notice", "image", "phone", "email", "date"} for name in names):
                counts["mismatch_with_site_features"] += 1
            else:
                counts["mismatch_without_site_features"] += 1
            for name in names:
                counts["mismatch_" + name] += 1


def main():
    if len(sys.argv) not in (3, 4) or (len(sys.argv) == 4 and sys.argv[3] != "--supported-only"):
        raise SystemExit("Usage: check-markdown-downstream.py <private.ndjson> <probe-binary> [--supported-only]")
    source, binary = sys.argv[1:3]
    supported_only = len(sys.argv) == 4
    counts = collections.Counter()
    page_equal = {}
    batch = []
    batch_bytes = 0
    with open(source, encoding="utf-8") as stream:
        for line in stream:
            case = json.loads(line)
            size = len(case["markdown"].encode())
            if size > 4_000_000:
                raise RuntimeError("A Markdown block exceeds the probe batch limit")
            if batch and (
                len(batch) >= 250
                or batch_bytes + size > 4_000_000
                or case["pre_class"] != batch[0]["pre_class"]
            ):
                compare(binary, batch, counts, page_equal, supported_only)
                batch = []
                batch_bytes = 0
            batch.append(case)
            batch_bytes += size
        compare(binary, batch, counts, page_equal, supported_only)
    print(
        json.dumps(
            {
                "snapshot_sha256": sha256(source),
                "probe_sha256": sha256(binary),
                "supported_only": supported_only,
                "pages": len(page_equal),
                "pages_exact": sum(page_equal.values()),
                "counts": dict(sorted(counts.items())),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: packages/core/rust/benchmarks/check-markdown-downstream.php
<?php

declare(strict_types=1);

use Symfony\Component\Process\Process;

require __DIR__.'/../../../../vendor/autoload.php';

$rawArguments = $_SERVER['argv'] ?? [];
if (! is_array($rawArguments)) {
    throw new InvalidArgumentException('Invalid command-line arguments.');
}

$arguments = [];
foreach ($rawArguments as $argument) {
    if (! is_string($argument)) {
        throw new InvalidArgumentException('Invalid command-line argument.');
    }

    $arguments[] = $argument;
}

if (! in_array(count($arguments), [3, 4], true) || (4 === count($arguments) && '--supported-only' !== $arguments[3])) {
    throw new InvalidArgumentException('Usage: php check-markdown-downstream.php <private.ndjson> <probe-binary> [--supported-only]');
}

/** @param array{markdown: string, php: string} $case
 * @return list<string>
 */
function features(array $case): array
{
    $source = $case['markdown'];
    $html = $case['php'];
    $checks = [
        'obfuscated_link' => 1 === preg_match('/#\[[^]]+\]\(/', $source),
        'notice' => str_contains($source, '> [!'),
        'image' => str_contains($source, '!['),
        'phone' => str_contains($html, 'data-rot="gry:'),
        'email' => str_contains($html, 'class=nojs'),
        'date' => str_contains($source, 'date('),
        'attributes' => 1 === preg_match('/\{[.#]|\{(?:id|class)=/', $source),
    ];

    $present = array_keys(array_filter($checks));

    return [] === $present ? ['unclassified'] : $present;
}

/**
 * @param list<array{page: string, markdown: string, pre_class: string, php: string}> $cases
 * @param array<string, int>                                                          $counts
 * @param array<string, bool>                                                         $pageEqual
 */
function compare(string $binary, array $cases, array &$counts, array &$pageEqual, bool $supportedOnly): void
{
    if ([] === $cases) {
        return;
    }

    $request = json_encode([
        'documents' => array_column($cases, 'markdown'),
        'fenced_code_pre_class' => $cases[0]['pre_class'],
        'supported_only' => $supportedOnly,
    ], \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR);
    $process = new Process([$binary]);
    $process->setInput($request);
    $process->setTimeout(null);
    $process->mustRun();

    $output = json_decode($process->getOutput(), true, flags: \JSON_THROW_ON_ERROR);
    if (! is_array($output) || count($output) !== count($cases)) {
        throw new RuntimeException('Probe returned a different document count.');
    }

    foreach ($cases as $index => $case) {
        $counts['blocks'] = ($counts['blocks'] ?? 0) + 1;
        $page = $case['page'];
        $pageEqual[$page] ??= true;
        $actual = $output[$index];
        if (null === $actual && $supportedOnly) {
            $counts['declined_to_php'] = ($counts['declined_to_php'] ?? 0) + 1;

            continue;
        }

        if ($actual === $case['php']) {
            $counts['exact'] = ($counts['exact'] ?? 0) + 1;
            if ($supportedOnly) {
                $counts['accepted_native'] = ($counts['accepted_native'] ?? 0) + 1;
            }
        } else {
            $counts['mismatch'] = ($counts['mismatch'] ?? 0) + 1;
            $pageEqual[$page] = false;
            $names = features($case);
            $siteFeatures = ['obfuscated_link', 'notice', 'image', 'phone', 'email', 'date'];
            $category = [] !== array_intersect($names, $siteFeatures) ? 'mismatch_with_site_features' : 'mismatch_without_site_features';
            $counts[$category] = ($counts[$category] ?? 0) + 1;
            foreach ($names as $name) {
                $key = 'mismatch_'.$name;
                $counts[$key] = ($counts[$key] ?? 0) + 1;
            }
        }
    }
}

$source = $arguments[1];
$binary = $arguments[2];
$supportedOnly = 4 === count($arguments);
$counts = [];
$pageEqual = [];
$batch = [];
$batchBytes = 0;
$stream = fopen($source, 'r');
if (false === $stream) {
    throw new RuntimeException('Cannot open snapshot: '.$source);
}

try {
    while (false !== $line = fgets($stream)) {
        $decoded = json_decode($line, true, flags: \JSON_THROW_ON_ERROR);
        if (! is_array($decoded) || ! is_string($decoded['page'] ?? null)
            || ! is_string($decoded['markdown'] ?? null) || ! is_string($decoded['pre_class'] ?? null)
            || ! is_string($decoded['php'] ?? null)) {
            throw new RuntimeException('Invalid Markdown snapshot record.');
        }

        $case = [
            'page' => $decoded['page'],
            'markdown' => $decoded['markdown'],
            'pre_class' => $decoded['pre_class'],
            'php' => $decoded['php'],
        ];
        $size = strlen($case['markdown']);
        if ($size > 4_000_000) {
            throw new RuntimeException('A Markdown block exceeds the probe batch limit.');
        }

        if ([] !== $batch && (count($batch) >= 250 || $batchBytes + $size > 4_000_000 || $case['pre_class'] !== $batch[0]['pre_class'])) {
            compare($binary, $batch, $counts, $pageEqual, $supportedOnly);
            $batch = [];
            $batchBytes = 0;
        }

        $batch[] = $case;
        $batchBytes += $size;
    }

    compare($binary, $batch, $counts, $pageEqual, $supportedOnly);
} finally {
    fclose($stream);
}

ksort($counts);
echo json_encode([
    'snapshot_sha256' => hash_file('sha256', $source),
    'probe_sha256' => hash_file('sha256', $binary),
    'supported_only' => $supportedOnly,
    'pages' => count($pageEqual),
    'pages_exact' => count(array_filter($pageEqual)),
    'counts' => (object) $counts,
], \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR)."\n";
